Correct and improve CI/CD setup documentation

rija opened the pull request:
# Pull request for issue: #1908

This is a pull request for the following functionalities:

* Update `docs/SETUP_CI_CD_PIPELINE.md` for accuracy and readability
* Create `docs/SETUP_PROVISIONING.md` to describe provisioning
* Updated documentation for AWS IAM custom policies to match current state 
* And make them JSON files the be easily read and edited in an IDE

## How to test?

### Pre-requisites:

* Create  a new regular user on AWS, make it part of the Developer group

* Create a new Gitlab project under `gigascience/Forks` group connected to your Github fork of Gigascience's gigadb-website repository
* Pick a new AWS region not yet used that's listed in the IAM policy docs `docs/awsdocs/policy-ec2.json`
* Request quota increase for the number of EIPs (5 -> 6) in the selected region (using https://eu-west-3.console.aws.amazon.com/servicequotas/home/requests, where `eu-west-3` is replaced by your selected region)

### Step by step instructions

Then diligently, from the perspective of a contractor or contributor, **not** from a code owner POV,  follow instructions in: 
1. `docs/SETUP_CI_CD_PIPELINE.md`
2. `docs/SETUP_PROVISIONING.md`

## How have functionalities been implemented?

After satisfying the pre-requisites, I have rebuilt Gitlab pipeline and AWS deployment from scratch, and updated the documentation along the way.

## Any issues with implementation?

N/a

## Any changes to automated tests?

Marked a functional as `@skip` as it is failing on develop branch

## Any changes to documentation?

Yes, audited and updated documentation related to CI/CD and provisioning

## Any technical debt repayment?

Updated documentation to reflect current usage

## Any improvements to CI/CD pipeline?

* Domain names for fork projects should all be subdomains to `gigadb.host`
* Use named profile `gigadb` for AWS credentials
* Make easier to debug in build and deploy jobs  issues related to gitlab variables capturing

rija committed:
tests: skip tests that fails on develop

they will be fixed in PR #1919

rija committed:
docs: update documentation for IAM custom policies

rija committed:
docs: update CI/CD setup docs's CI part with corrections

rija committed:
docs: Improve CI/CD pipeline setup

Add missing variables needed for provisioning
Reorder section to be more coherent
introduce new protocol for domain names

rija committed:
fix: improve debugging of build stage

rija committed:
fix: in gigadb build jobs config use tee instead of double chevron

as double chevron return an error status is it is sent empty data

rija committed:
fix: in gigadb deploy job use tee instead of double chevron

rija committed:
fix: in gigadb deploy job remove two debug lines

rija committed:
fix: output application log for debugging

rija committed:
docs: add more needed variables,explain pipeline/provisioning connection

rija committed:
docs: correct errors in the docs and missing info re AWS

rija committed:
docs: documents vars needed for staging gigareview and acceptance tests

rija committed:
feat: use AWS named profile instead of default

rija committed:
docs: move provisioning in its own docs and correct its errors

rija committed:
test: skip failing test on develop

rija committed:
test: remove a skip as the issue was fixed upstream

rija committed:
docs: move optional section to the end

rija committed:
docs: make all custom IAM policies JSON

rija committed:
docs: add more info on gitlab variables for groups, projects, tokens

rija committed:
docs: shorten lines that are not tables or lists

rija committed:
docs: Add a CHANGELOG entry

rija committed:
fix: accidentally commit a change not meant to be commited

rija committed:
docs: remove redundant instructions

rija committed:
docs: update policy docs to add more regions

rija committed:
docs: add a preamble to provisioning docs and updated aws cli doc

pli888 commented on docs/SETUP_CI_CD_PIPELINE.md:
```diff
@@ -64,6 +64,97 @@ others in the Forks group.
+| live | production | the real live web site product hosted on AWS cloud |
+
+The local environment is on the developer's machine, that's what the dev environment refers to. 
+The CI environment is implictely created by the CI part of GitLab, that's where the code is deployed for the execution of the automated tests.
```
Typo in `implictely` ~> `implicitly`.

pli888 commented on docs/SETUP_CI_CD_PIPELINE.md:
```diff
@@ -64,6 +64,97 @@ others in the Forks group.
+
+A functionality of Gitlab is to store environment variables, so that we can use them in our deployed applications.
+Because there is multiple deployment environments and the variables often differ from one to the other, 
+Gitlab variables can be categorised and into different environment which are:
```
Gitlab variables can be categorised into different environments which are:

pli888 commented on docs/SETUP_CI_CD_PIPELINE.md:
```diff
@@ -64,6 +64,97 @@ others in the Forks group.
+
+The environment is associated upon creation to each variable as one of its attributes.
+
+>**Note:** there is no connection between the environment of variables and the deployment environments listed in previous section. 
```
It feels like there is a connection between the environment of variables and deployment environments based on the following sentence?

pli888 commented on docs/SETUP_CI_CD_PIPELINE.md:
```diff
@@ -64,6 +64,97 @@ others in the Forks group.
+
+Furthermore, variables have a hierarchal organisation that map to groups and projects.
+So that, when there are variables is needed by all developers with the same value, such variable can be defined in a parent group, 
+which allow and sub-groups or sub-projects to access it without having to define it.
```
Delete `and`.

pli888 commented on docs/SETUP_CI_CD_PIPELINE.md:
```diff
@@ -64,6 +64,97 @@ others in the Forks group.
+So that, when there are variables is needed by all developers with the same value, such variable can be defined in a parent group, 
+which allow and sub-groups or sub-projects to access it without having to define it.
+
+##### Group and projects The diagram below show the hierarchy we have in place.
```
`The diagram below show the hierarchy we have in place.` should be in a separate line below, not part of the sub-title.

pli888 commented on docs/SETUP_CI_CD_PIPELINE.md:
````diff
@@ -64,6 +64,97 @@ others in the Forks group.
+    └── gigadb-website
+```
+
+All the leaves op the tree are _Gitlab projects_, the nodes are _Gitlab groups_.
````
`op` should be `of`.

pli888 commented on docs/SETUP_CI_CD_PIPELINE.md:
```diff
@@ -161,34 +249,114 @@ They have two very important use:
 us to store variables and specify for which environment this variable is bound to (for organisation and security).
  * Gitlab stages and jobs must be tied to a specific environment, so that pipelines don't leak variables.
 
+#### Create DNS record for accessing endpoint on staging and on live servers
```
Maybe this section would be better placed after the `AWS dashboard` section because DNS A records cannot be created with knowing the elastic IP addresses?

pli888 commented on docs/SETUP_CI_CD_PIPELINE.md:
```diff
@@ -161,34 +249,114 @@ They have two very important use:
+
+1. For Elastic IPS: ``eip-<application>-<environment>[-<sub-system>]-<IAM Role Username>``, e.g: ``eip-gigadb-staging-John`` or ``eip-gigadb-files-staging-John``
+1. For SSH Key pair: ``aws-<application>-<AWS region>-<IAM Role Username>.pem``, e.g: ``aws-gigadb-eu-north-1-John.pem``
+The private part of the SSH Key pair needs to be dowloaded to your developer machine in the ``~/.ssh`` directory and with 
```
Typo `dowloaded`

pli888 commented on docs/SETUP_CI_CD_PIPELINE.md:
```diff
@@ -161,34 +249,114 @@ They have two very important use:
+| REMOTE_SMTP_PASSWORD | SMTP password |
+| REMOTE_SMTP_PORT | 563 |
+| REMOTE_SMTP_USERNAME | SMTP username |
+| gigadb_db_host | keep empty, it will be overwritten by the provising script with RDS endpoint |
```
`provising` should be `provisioning`

pli888 commented on docs/SETUP_CI_CD_PIPELINE.md:
```diff
@@ -161,34 +249,114 @@ They have two very important use:
+| PORTAINER_PASSWORD | Pick a password |
+| remote_fileserver_hostname | files.yoursubdomain.gigadb.host | 
+| REMOTE_PUBLIC_HTTP_PORT | 80 |
+| REMOTE_PUBLIC_HTTPS_PORT | 443 |
```
Can delete these `REMOTE_PUBLIC_HTTP_PORT` and `REMOTE_PUBLIC_HTTPS_PORT` rows because they have already been defined above on lines 335 and 336.

pli888 commented on docs/SETUP_CI_CD_PIPELINE.md:
```diff
@@ -222,44 +415,29 @@ to depart from the default.
 
-> Ensure the value of ``environment:name:`` in GitLab configuration matches the environment that you have created in Gitlab dashboard under ``Deployments > Environments``
+>Ensure the value of ``environment:name:`` in GitLab configuration matches the environment that 
+>you have created in Gitlab dashboard under ``Deployments > Environments``
```
It is `Operate > Environments` in the current GitLab web UI.

pli888 commented on docs/SETUP_PROVISIONING.md:
```diff
@@ -0,0 +1,1505 @@
+## Relationship between Gitlab pipeline and provisioning
+
+The continuous integration part of the Gitlab pipeline is not dependent on the AWS deployed infrastructure and can be run anytime.
+The provisioning of a new AWS infrastructure and deployment of Gigadb applicaiton onto it requires 
```
`applicaiton` should be `application`.

pli888 commented on docs/SETUP_PROVISIONING.md:
```diff
@@ -0,0 +1,1505 @@
+
+| phase | description | operator actions | required Gitlab pipeline state |
+| -- | -- | -- | -- |
+| 1.TF init | The local terraform state is initiliased from a remote Gitlab state (which is created on the fly if it doesn't exist) | runs `tf_init.sh` | no state required |
```
`initiliased` should be `initialised`.

pli888 commented on docs/SETUP_PROVISIONING.md:
```diff
@@ -0,0 +1,1505 @@
+
+>**Note:** You cannot run the build stage until you have run `ansible_init.sh`, because the build stage needs AWS resources details uploaded by that script
+
+>**Note:** The above implies that you cannot run the Ansible playbooks until you have run the build stage of the Gitlab pipeline, 
```
Delete comma.

pli888 commented on docs/SETUP_PROVISIONING.md:
```diff
@@ -0,0 +1,1505 @@
+
+>**Note:** The above implies that you cannot run the Ansible playbooks until you have run the build stage of the Gitlab pipeline, 
+
+>**Note:** GigaDB application is made of web-based and backoffice subsystems. Their deployment requires the build stage of the Gitlab pipeline to have been peformed frist.
```
`peformed frist` should be `performed first`.

pli888 commented on docs/SETUP_PROVISIONING.md:
```diff
@@ -0,0 +1,1505 @@
+### Pre-requisites: Setup pipeline and Install the tools
+
+1. You have implemented the instructions in `docs/SETUP_CI_CD_PIPELINE.md`
+2. Install AWS CLI by following AWS's instructions[here](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html).
```
Add space into `instructions[here]`.

pli888 commented on docs/SETUP_PROVISIONING.md:
```diff
@@ -0,0 +1,1505 @@
+1. You have implemented the instructions in `docs/SETUP_CI_CD_PIPELINE.md`
+2. Install AWS CLI by following AWS's instructions[here](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html).
+3. Install Terraform with `brew install terraform`
+4. Install Ansible with `brew install ansible`
```
Might be worth adding a note to say ansible version 2.16 is required. @alli83 said she had to downgrade from version 2.17.

pli888 commented on docs/SETUP_PROVISIONING.md:
```diff
@@ -0,0 +1,1505 @@
+You can now start a Gitlab pipeline for the git branch or tag you want to deploy (the main branch is called `develop`).
+
+
+#### 3. Configure the infastructure
```
Typo in `infastructure`.

pli888 commented on docs/SETUP_PROVISIONING.md:
```diff
@@ -0,0 +1,1505 @@
+
+The application should be available at the url defined in $REMOTE_HOME_URL for the staging environment
+
+For the FTP, trigger on the Gitlab pipelione the build job `PureFtpdBuildStaging`, and upon success trigger the deployment with `PureFtpdDeployStaging`.
```
Typo in `pipelione`.

pli888 commented on docs/SETUP_PROVISIONING.md:
```diff
@@ -0,0 +1,1505 @@
+>**Note:** The `build_staging` and `sd_gigadb` jobs are automated, so if you take too long to perform step 3, the jobs may run and fails.
+>That's fine, you can just trigger the jobs manually whenever you have completed step 3.
+
+#### 5. Configure the bastion server and deploy the back-office subsystems
```
I needed to have bastion server already provisioned by its Ansible playbook otherwise I get this error when running `sd_gigadb` in CI pipeline:
```
[21-Jul-2024 20:46:55 UTC] PHP Warning:  Unknown: php_network_getaddresses:
getaddrinfo failed: Name or service not known in Unknown on line 0. Running Yii command generatefiletypes.
CDbException: The table "file_type" for active record class "FileType" cannot be found in the database. in /var/www/vendor/yiisoft/yii/framework/db/ar/CActiveRecord.php:2391
```

Might be better to incorporate this part into section 3.

rija replied:
Hi @pli888,
 
the problem  here is that are there parts of the bastion playbook that requires the `build_staging` and `sd_gigadb` to have run beforhand.

from the error you get it sounds like there is a  network  setup done in bastion playbook that should have been done in bootstrap playbook (run by `ansible_init.sh`.

I'll create a ticket for that. In the meantime I'll make the  incantation to bastion playbook to appear in both section.

rija replied:
Created a ticket #1985 for the above issue.

rija replied:
Hi @pli888,

Instead of running the `bastion_playbook.yml` twice, we can use the `--tags` to control what tasks need to run when:

Before running the gitlab pipeline: 

env OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES ansible-playbook -i ../../inventories bastion_playbook.yml -e "gigadb_env=staging" --tags="fix-centos-eol-issues,setup-fail2ban,setup-docker-ce,restore-db-on-rds, create-metadata-bucket-credentials-file,setup-uploading-db-dump-to-s3,rclone-tool,node-exporter"


After running the gitlab pipeline:

env OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES ansible-playbook -i ../../inventories bastion_playbook.yml -e "gigadb_env=staging" --tags="files-url-updater,readme-tool,excel-spreadsheet-uploader,files-metadata-tool,load-latest-db"


I'll test those work and update the doc accordingly. 
It also offers us a strong hint on how to implement #1985

pli888 commented on docs/SETUP_PROVISIONING.md:
```diff
@@ -0,0 +1,1505 @@
+
+### Setting up your Live environment
+
+After you have setup your staging environmnet, you can optionally set up a live environment.
```
`environmnet` should be `environment`.

pli888 commented on docs/SETUP_PROVISIONING.md:
```diff
@@ -0,0 +1,1505 @@
+### Setting up your Live environment
+
+After you have setup your staging environmnet, you can optionally set up a live environment.
+In general this is not necessary, unless you want to test changes to infrastructure, Gitlabp pipeline or provisioning scripts.
```
Typo in `Gitlabp`

pli888 commented on docs/SETUP_PROVISIONING.md:
```diff
@@ -0,0 +1,1505 @@
+
+1. You have implemented the instructions in `docs/SETUP_CI_CD_PIPELINE.md`
+2. Install AWS CLI by following AWS's instructions[here](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html).
+3. Install Terraform with `brew install terraform`
```
**note**: I'm using Terraform 1.8.5

pli888 commented on docs/SETUP_PROVISIONING.md:
```diff
@@ -0,0 +1,1505 @@
+Actual changes to the infrastructure are performed with the command ``terraform apply``.
+
+The state file Terraform works with can be a local file (ending with suffix .tfstate), but it is recommended 
+to use a remote state backend wich has the following benefits:
```
`wich` should `which`

pli888 commented on docs/SETUP_PROVISIONING.md:
```diff
@@ -0,0 +1,1505 @@
+| ansible.properties | created by ``ansible_init.sh``, holds variables assignment used by Ansible to configure deployed application| ansible-playbook | 
+| getIAMUserNameToJSON.sh | copied by ``ansible_init.sh`` | terraform |
+| output/ | created by ``ansible-playbook`` to store a copy of Docker certs | docker |
+| (webapp|files|users|bastion)_playbook.yml | copied by ``ansible_init.sh`` | ansible-playbook | 
```
`(webapp|files|users|bastion)` doesn't seem to be rendered properly in markdown viewer in PHPStorm.

rija replied:
Yes, and there was one missing (`bootstrap_playbook.yml`)

pli888 commented on docs/SETUP_PROVISIONING.md:
````diff
@@ -0,0 +1,1505 @@
+
+Here are the steps:
+```
+# To check whether bastion server is accessible by loggin in
````
`loggin` should be `logging`.

pli888 commented on docs/SETUP_PROVISIONING.md:
```diff
@@ -0,0 +1,1505 @@
+GitLab. Check them on the project environment variables page after the Ansible
+provisioning has completed. This is done by the `docker-postinstall` role.
+
+| Variable's Key | Environnments | Description |
```
Typo in `Environnments`.

pli888 commented on docs/SETUP_PROVISIONING.md:
```diff
@@ -0,0 +1,1505 @@
+
+The bastion playbook will create a `gigadb` database containing data from
+`sql/production_like.pgdmp`. This file is created by running `./up.sh` when
+spinning up a local GigaDB on your development platform.
```
I think the bastion playbook will create a database on RDS by restoring from a backup from S3 now.

pli888 commented on docs/SETUP_PROVISIONING.md:
````diff
@@ -0,0 +1,1505 @@
+- docker-compose --tlsverify -H=$REMOTE_WEBAPP_DOCKER -f ops/deployment/docker-compose.production-envs.yml exec -T web /usr/local/bin/enable_sites gigadb.$GIGADB_ENV.https
+```
+
+When the container service ``web`` is started, it will also execute that ascript to enable the default http configuration for Gigadb.org,
````
`ascript` should be `script`.

pli888 commented on docs/SETUP_PROVISIONING.md:
```diff
@@ -0,0 +1,1505 @@
+The TLS certificate for the domain $REMOTE_HOSTNAME are managed by the script ``ops/scripts/setup_cert.sh``
+
+It will create new certificates with Let's encrypt if none exists on the target deployment or remotely as GitLab variables.
+if existing on the target deployment, the script will request renewal which Let's Encrypt will only perform if close to expiration.
```
`if` should be `If`.

pli888 commented on docs/SETUP_PROVISIONING.md:
```diff
@@ -0,0 +1,1505 @@
+| tls_chain_pem | live | contains the additional intermediate certificate or certificates that web browsers will need in order to validate the server certificate |
+
+When provisioning a new staging or live environment, ``ops/scripts/setup_cert.sh`` will attempt to pull these variables if they exist.
+This way, we won't unnecessarily creat certificate everytime we need to re-create a staging and deployment, 
```
`creat` should be `create`.

pli888 commented on docs/SETUP_PROVISIONING.md:
```diff
@@ -0,0 +1,1505 @@
+console for the RDS service in the `Log & events` tab.
+
+
+### Tools
```
Can probably delete everything from here to the end of this file because this part of the documentation has been covered above?

pli888 commented on docs/SETUP_PROVISIONING.md:
````diff
@@ -0,0 +1,1505 @@
+
+```
+$ cd ops/infrastructure/envs/staging
+$ terraform destroy
````
Should be `AWS_PROFILE=gigadb terraform destroy`?

pli888 commented on docs/SETUP_PROVISIONING.md:
````diff
@@ -0,0 +1,1505 @@
+
+```
+$ cd ops/infrastructure/envs/live
+$ terraform destroy
````
`AWS_PROFILE=gigadb terraform destroy`?

pli888 commented on docs/SETUP_PROVISIONING.md:
```diff
@@ -0,0 +1,1505 @@
+
+##### How to manually ssh to dockerhost through the bastion for debugging purpose
+Sometimes, it would be useful to log into dockerhost server manually for debugging. There are two important points to keep in mind:
+1. Get the public DNS or private ip address of web server or files server from `terraform output` or EC2 dashboard. 
```
Probably best to use `AWS_PROFILE=gigadb terraform output` here?

pli888 reviewed (changes requested):
**praise**: Following the updated documentation, I was able to deploy a GigaDB staging environment in ap-south-1 region using a new AWS IAM user account from a new fork project on Gitlab.
**thought**: Regarding the skipped tests, maybe they are not required to be skipped as all tests are passing in the upstream develop at the moment.
**chore**: Rebase and fix conflict in CHANGELOG.md.
**chore**: Address comments and fix typos in documentation.
**chore**: The **Mirroring your forked gigadb-website repository from GitHub** can be updated due to changes in GitLab web UI. To connect a Github repo, I did this:

>Click on blue `New project` button followed by `Import project` link. On the Import project page, click on `Repository by URL` button and enter the GitHub repository URL, e.g. https://github.com/pli888/gigadb-website. Check the `Mirror repository` checkbox and check `Public` visibility Level option. Finally click blue `Create project` button.

**chore**: I found these issues in the **Configuring your GitLab gigadb-website project** section:

> The https://gitlab.com/gigascience/forks/pli888-gigadb-website/settings/repository link is wrong, should be https://gitlab.com/gigascience/forks/pli-gigadb-website/-/settings/

> The https://gitlab.com/gigascience/forks/pli888-gigadb-website/settings/ci_cd link is broken, should be https://gitlab.com/gigascience/forks/pli-gigadb-website/-/settings/ci_cd

> There doesn't seem to be a Test coverage parsing text field anymore so this text can be deleted: `The Test coverage parsing text field should also contain: \ \ Lines:\s*(\d+.\d+\%). Click on the Save changes green button.`

> Need to create `dev` environment via Operate > Environments so these environments are available to select when creating variables

kencho51 reviewed (commented):
Hi @rija 

I think the following steps in `SETUP_CI_CD_PIPELINE.md` is out of date:
```
### Mirroring your forked gigadb-website repository from GitHub

To begin, mirror your forked GitHub gigadb-website repository as a GitLab 
project. This is done by adding your GitHub gigadb-website repository to the 
GitLab Gigascience Forks organisation. To do this:

* Log into GitLab and go to the 
[gigascience/Forks page](https://gitlab.com/gigascience/forks).
 
* Click on *New Project* followed by *CI/CD for external repo* and then 
*GitHub*. 

* On the *Connect repositories from GitHub page*, click on the 
*List your GitHub repositories* green button. Find the repository fork of 
`gigadb-website` that you want to perform CI/CD on.

* Under the *To GitLab* column, select *gigascience/forks* to connect your repo 
to this GitLab group. Also, provide a name for the repo, *e.g.* 
`pli888-gigadb-website` so that you can differentiate your repository from 
others in the Forks group.

* Click the *Connect* button to create the mirror of the GitHub repository on
GitLab.
```

Because I could not create a new gitlab repo by following the above steps, the information shows in the [gitlab fork ](https://gitlab.com/gigascience/forks) is different.

rija committed:
fix: explicitly log in to gitlab

rija committed:
test: restore skipped test for md5 update as ticket 1919 has been merged

rija committed:
docs: Update the Mirroring your forked gigadb-website repository section

rija committed:
docs: remove redundant variables

rija committed:
docs: change Deployments  to Operate

rija committed:
docs: applicaiton to application

rija committed:
docs: initiliased to initialised

rija committed:
docs: delete trailing comma in note section

rija committed:
docs: peformed frist to performed first

rija committed:
docs: add space between instructions and here

rija committed:
docs: add note about Ansible version

rija committed:
docs: typo in infastructure and missing AWS_PROFILE prefix

rija committed:
docs: correct typos in pipelione

rija committed:
docs: run bastion playbook twice due to a chicken and egg problem

rija committed:
docs: correct typo in environmnet

rija committed:
docs: correct typo in Gitlabp

rija committed:
docs: correct a typo in wich

rija committed:
docs: reformat the list of files associated with ansible

rija committed:
docs: correct typo in loggin

rija committed:
docs: correct typo in Environnments

rija committed:
docs: Update how the non-live environments get their DB from

rija committed:
docs: correct typo in ascript

rija committed:
docs: corrects lowercase if

rija committed:
docs: correct typo in creat

rija committed:
docs: add AWS_PROFILE prefix

rija committed:
docs: remove duplicated content

rija committed:
docs: correct typo in implictely

rija committed:
docs: remove extra and

rija committed:
docs: correct the typo in provising

rija committed:
docs: correct typo in op

rija committed:
docs: move the DNS record section after the AWS dashboard section

rija committed:
docs: correct typo in dowloaded

rija committed:
docs: correct title formatting for hierarchy diagram section

rija committed:
docs: remove incorrect statement about environments

rija committed:
docs: correct typo in infastructure

rija committed:
docs: add missing AWS_PROFILE prefix

rija commented:
> Hi @rija
> 
> I think the following steps in `SETUP_CI_CD_PIPELINE.md` is out of date:
> 
> ```
> ### Mirroring your forked gigadb-website repository from GitHub
> 
> To begin, mirror your forked GitHub gigadb-website repository as a GitLab 
> project. This is done by adding your GitHub gigadb-website repository to the 
> GitLab Gigascience Forks organisation. To do this:
> 
> * Log into GitLab and go to the 
> [gigascience/Forks page](https://gitlab.com/gigascience/forks).
>  
> * Click on *New Project* followed by *CI/CD for external repo* and then 
> *GitHub*. 
> 
> * On the *Connect repositories from GitHub page*, click on the 
> *List your GitHub repositories* green button. Find the repository fork of 
> `gigadb-website` that you want to perform CI/CD on.
> 
> * Under the *To GitLab* column, select *gigascience/forks* to connect your repo 
> to this GitLab group. Also, provide a name for the repo, *e.g.* 
> `pli888-gigadb-website` so that you can differentiate your repository from 
> others in the Forks group.
> 
> * Click the *Connect* button to create the mirror of the GitHub repository on
> GitLab.
> ```
> 
> Because I could not create a new gitlab repo by following the above steps, the information shows in the [gitlab fork ](https://gitlab.com/gigascience/forks) is different.

Hi @kencho51,
I've updated that section of the documentation. you can try again.
I've also added information regarding installing the correct version of Ansible, you can try out.

rija committed:
test: explicit docker login in each test job

rija commented:
Hi @pli888,
>thought: Regarding the skipped tests, maybe they are not required to be skipped as all tests are passing in the upstream develop at the moment.

I've removed the one '@skip' mention explicitly related to the PR #1919.
I've left over '@skip' mentions associated with anonymous TODO despite linked to a recent change from the same PR (use of RDS automated backup) because those tests are likely to break as they were not updated in that PR and seem to be still testing the old way of restoring DB

rija committed:
docs: correct broken gitlab setting links

rija committed:
docs: add a note that all environments (incl. dev) need to be created

rija committed:
docs: remove the redundant instruction about test coverage parsing text

rija commented:
Hi @pli888,

>**chore**: I found these issues in the **Configuring your GitLab gigadb-website project** section:

> The https://gitlab.com/gigascience/forks/pli888-gigadb-website/settings/repository link is wrong, should be https://gitlab.com/gigascience/forks/pli-gigadb-website/-/settings/

> The https://gitlab.com/gigascience/forks/pli888-gigadb-website/settings/ci_cd link is broken, should be https://gitlab.com/gigascience/forks/pli-gigadb-website/-/settings/ci_cd

> There doesn't seem to be a Test coverage parsing text field anymore so this text can be deleted: `The Test coverage parsing text field should also contain: \ \ Lines:\s*(\d+.\d+\%). Click on the Save changes green button.`

> Need to create `dev` environment via Operate > Environments so these environments are available to select when creating variables

Made corrections regarding all those points.

rija committed:
test: need explicit login in gigadb tools gitlab config too

rija committed:
revert: these changes will be done on another branch

there are needed for blue/green infrastructure deployment
so they more urgent than this branch

pli888 reviewed (approved):
With the changes made to the documentation and because I already was able to deploy a staging server using a new AWS IAM user and new Gitlab gigadb-website project using it, I am happy to approve this PR.

kencho51 reviewed (approved):
Hi @rija,

**praise:** I was able to spin up gigadb website from `eu-north-1` region and new gitlab fork repo
**praise:** The `docs/SETUP_CI_CD_PIPELINE.md` and `docs/SETUP_PROVISIONING.md` docs provide complete guide in setting up gigadb website as a new user
**issue(non-blocking):** It is a know issue that ansible (with version > 2.17.x) will cause `/bin/sh: /usr/bin/python3: No such file or directory\r\n` when executing the playbooks, details refer to https://github.com/gigascience/gigadb-website/issues/1986
**issue(non-blocking):** Occasionally, when executing the webapp playbook or the bastion playbook, ansible will throw this error `Failed to download metadata for repo 'appstream': Cannot prepare internal mirrorlist: No URLs in mirrorlist`, but this has been addressed in https://github.com/gigascience/gigadb-website/issues/1897, and will be fixed when we switch to Centos Stream 9
**issue(non-blocking):** The `brew install ansible@9` with version `2.16.8` would cause `[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self-signed certificate in certificate chain (_ssl.c:1000)` on my dev, the reason may be rooted from here: https://stackoverflow.com/questions/63534262/how-to-fix-following-ansible-galaxy-ssl-error, my temp fix to install ansible 2.15.2 with pip which allows specify the version
**suggestion(non-blocking):** Agree to have to have a table showing which aws region is being used by which developer, and which region is available for new development.

Since there is no any blockers which seems only happened to me and I was able to fix them, so I would like to approve this PR.

rija committed:
docs: move AWS sect. from pipeline to provisioning doc, add AWS regions

rija committed:
docs: add missing step to Ansible install

rija committed:
docs: add a note regarding gitlab variables for domain names

rija commented:
Hi @pli888, @kencho51,

I've made updates to this PR with following changes:

* Correct erroneous information about deployment to staging and live from Gitlab pipeline
* Move the AWS sections from the pipeline doc to the top of the provisioning doc as it is more relevant in the latter doc
* In the provisioning doc,  add a section to list AWS regions in a table and indicates which ones are free and which  one are used 
* In the provisioning doc, amended the install instructions for Ansible to add the missing information

I won't add anything more to this PR, so once you've reviewed it and it's approved again, I'll merge it.

kencho51 commented on docs/SETUP_PROVISIONING.md:
```diff
@@ -0,0 +1,698 @@
+|  ap-southeast-1 | Singapore | gigabyte, gigablog | |
+|  ap-southeast-2 |  Sidney | gigadb (alternative) | |
+| ap-southeast-4 | Melbourne | | Under preparation, available soon |
+| ap-northeast-2 | Osaka | | Don't use, it doesn't have Centos 8 Stream X86_64 |
+| ap-northeast-3 | Seoul | |Don't use, it doesn't have Centos 8 Stream X86_64|
```
Hi @rija:

`ap-northeast-2` is used by me, and its region is Seoul.
`ap-northeast-3` region should be Osaka.

Details at here: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-regions-availability-zones.html

And regarding the current ami owner `125523088429` we are using, who does provide `Centos 8 Stream X86_64` ami in the Seoul region as you can see from this query: https://ap-northeast-2.console.aws.amazon.com/ec2/home?region=ap-northeast-2#Images:visibility=public-images;owner=125523088429;imageName=:CentOS%20Stream%208%20x86_64;v=3;$case=tags:false%5C,client:false;$regex=tags:false%5C,client:false

But the `Centos 8 Stream X86_64` ami  is unavailable in the Osaka (ap-northeast-3) region.

FYI, I was able to deploy alternative gigadb website from Europe (Stockholm) `eu-north-1`.  

In fact, I was wondering is it necessary to mention `Don't use, ...` in the comments? If the region cannot be used, then is it better not to mention it in the doc?

rija replied:
Thanks @kencho51, 
I've updated the table accordingly.
I've also removed Spain and Melbourne for now as they don't have the AMI we use currently.

rija committed:
docs: correct the AWS region table based on feedback

kencho51 reviewed (approved):
Hi @rija,

**praise:** The aws region table in the doc is cleaner and easy to read now.
**praise:** Further information on how to use brew to install ansible earlier version.

Happy to approve again.

rija committed:
docs: fix typo

pli888 reviewed (approved):
**praise**: Changes relating to AWS regions and Ansible installation look good to me.
**note**: We should remember to clean up the AWS resources and Gitlab Fork projects that were used to review this PR.

Happy to approve this PR.
